src/shims/env.rs

```diff
@@ -123,30 +142,31 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         })
     }
 
-    fn getenvironmentstringsw(&mut self) -> InterpResult<'tcx, Scalar<Tag>> {
+    fn GetEnvironmentStringsW(&mut self) -> InterpResult<'tcx, Scalar<Tag>> {
         let this = self.eval_context_mut();
         this.assert_target_os("windows", "GetEnvironmentStringsW");
 
         // Info on layout of environment blocks in Windows: 
         // https://docs.microsoft.com/en-us/windows/win32/procthread/environment-variables
         let mut env_vars = std::ffi::OsString::new();
         for &item in this.machine.env_vars.values()? {
-            let env_var = this.read_os_str_from_target_str(Scalar::from(item))?;
+            let env_var = this.read_os_str_from_wide_str(Scalar::from(item))?;
             env_vars.push(env_var);
             env_vars.push("\0");
         }
+        // Allocate environment block & Store environment variables to environment block.
+        // Final null terminator(block terminator) is added by `alloc_os_str_to_wide_str`.
+        let envblock_ptr = this.alloc_os_str_as_wide_str(&env_vars, MiriMemoryKind::WinHeap.into());
```

Comment:
This is not technically a `WinHeap` allocation... I think `Machine` is a better memory type for this.

Reply:
I tried to replace `MiriMemoryKind::WinHeap` with `MiriMemoryKind::Machine`, but MIRI gives me a UB: Stacked Borrows error.

```
error: Undefined Behavior: no item granting read access to tag <untagged> found in borrow stack.
  --> C:\Users\lonelyjoe\.rustup\toolchains\miri\lib\rustlib\src\rust\src\libstd\sys\windows\os.rs:99:23
   |
99 |                 while *p.offset(len) != 0 {
   |                       ^^^^^^^^^^^^^^ no item granting read access to tag <untagged> found in borrow stack.
   |
   = help: this indicates a potential bug in the program: it performed an invalid operation, but the rules it violated are still experimental
   = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
   = note: inside call to `<std::sys::windows::os::Env as std::iter::Iterator>::next` at C:\Users\lonelyjoe\.rustup\toolchains\miri\lib\rustlib\src\rust\src\libstd\env.rs:172:9
   = note: inside call to `<std::env::VarsOs as std::iter::Iterator>::next` at C:\Users\lonelyjoe\.rustup\toolchains\miri\lib\rustlib\src\rust\src\libstd\env.rs:154:9
   = note: inside call to `<std::env::Vars as std::iter::Iterator>::next` at C:\Users\lonelyjoe\.rustup\toolchains\miri\lib\rustlib\src\rust\src\liballoc\vec.rs:2015:32
   = note: inside call to `<std::vec::Vec<(std::string::String, std::string::String)> as std::vec::SpecExtend<(std::string::String, std::string::String), std::env::Vars>>::from_iter` at C:\Users\lonelyjoe\.rustup\toolchains\miri\lib\rustlib\src\rust\src\liballoc\vec.rs:1927:9
   = note: inside call to `<std::vec::Vec<(std::string::String, std::string::String)> as std::iter::FromIterator<(std::string::String, std::string::String)>>::from_iter::<std::env::Vars>` at C:\Users\lonelyjoe\.rustup\toolchains\miri\lib\rustlib\src\rust\src\libcore\iter\traits\iterator.rs:1660:9
note: inside call to `<std::env::Vars as std::iter::Iterator>::collect::<std::vec::Vec<(std::string::String, std::string::String)>>` at tests/run-pass/env.rs:15:23
  --> tests/run-pass/env.rs:15:23
   |
15 |     println!("{:#?}", env::vars().collect::<Vec<_>>());
```

Reply:
Ah, that's interesting... might actually be a bug in libstd. I'll have to investigate.

Reply:
Indeed this is a libstd bug: https://github.com/rust-lang/rust/pull/70479
I am fine landing this with `WinHeap` for now, if you add a link to that PR and a note to change this to `Machine` once the PR lands in Rust.